Add documentation for merge on red, build analysis, and v1/v2 tests (#84982)
dotnet/runtime: docs/workflow/ci/failure-analysis.md

# Analyzing Failures with Build Analysis and Known Issues

> 

Reviewer: Add a table of contents here.

    - The field `ExcludeConsoleLog` should be `true` as much as possible. Setting it to false will mean all failures appear in the same log and since you didn't exclude the overall logs, all failures in the collection will get attributed to this one's issue bucket and errors will not be investigated properly. Due to limitations in Known Issues around rate limiting and xUnit resiliency, setting `ExcludeConsoleLog=false` is necessary in two scenarios:
      + Nested tests as reported to Azure DevOps. Essentially this means theory failures, which look like this when reported in Azure DevOps: ![xUnit theory seen in azure devops](theory-azdo.png).
        Adding support for this requires too many API calls, so using the console log here is necessary.
      + Native crashes in libraries also require using the console log. This is needed as the crash corrupts the test results to be reported to Azure DevOps, so only the console logs are left.
    - Optionally you can add specifics as needed like leg, configuration parameters, available dump links.

> Once the issue is open, feel free to rerun the `Build Analysis` check and the issue should be recognized as know if all was filed correctly. there are some known limitations as previously described. It also has no support at looking at fields outside the error message, the stacktrace, and the console log in the helix queue, so this approach doesn't yet cover devices too well yet.

Reviewer: I would rewrite "...feel free to rerun the `Build Analysis` check and the issue should be recognized as know if all was filed correctly. there are some known limitations as previously described" as:
"...feel free to rerun the `Build Analysis` check and the issue you just filed should be recognized as known, if everything was done correctly. However, it is important to remember there are some known limitations in this system, as previously described"

After you do this, if the failure is occuring frequently as per the data captured in the recently opened issue, please disable the failing test(s) with the corresponding issue link tracking the disable in a follow-up Pull Request

* Update the tracking issue with the label `disabled-test`.
* For libraries tests add a [`[ActiveIssue(link)]`](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.XUnitExtensions/src/Attributes/ActiveIssueAttribute.cs) attribute on the test method. You can narrow the disabling down to runtime variant, flavor, and platform. For an example see [File_AppendAllLinesAsync_Encoded](https://github.com/dotnet/runtime/blob/cf49643711ad8aa4685a8054286c1348cef6e1d8/src/libraries/System.IO.FileSystem/tests/File/AppendAsync.cs#L74)
* For runtime tests found under `src/tests`, please edit [`issues.targets`](https://github.com/dotnet/runtime/blob/main/src/tests/issues.targets). There are several groups for different types of disable (mono vs. coreclr, different platforms, different scenarios). Add the folder containing the test and issue mimicking any of the samples in the file.

> 

Reviewer: Should we specify here whether they should disable the test in their same PR, or a separate one?

Author: The text specifies: 

> please disable the failing test(s) with the corresponding issue link tracking the disable *in a follow-up Pull Request*

It's probably better to be able to revert the one commit in the issue solution PR.

Reviewer: Oops my bad, I didn't realize at the time it was specified above. Thanks for the clarification Juan

Validation may fail for several reasons, and for each one we have a different recommended action:

### Option 1: You have a defect in your PR

* Simply push the fix to your PR branch, and validation will start over.

### Option 2: There is a flaky test that is not related to your PR

* Your assumption should be that a failed test indicates a problem in your PR. (If we don't operate this way, chaos ensues.) However, there's often subtle regressions and flaky bugs that might have slipped into the target branch.
  * Reruns might help, but we tend to be conservative with them as they tend to spike our resource usage. Opt to use them only if there's no known issue that can be correlated to the failures and it's not clear if the errors could be correlated. Try to rerun only the particular legs if possible by navigating to the GitHub Checks tab choose "re-run failed checks".

>   * There's the possibility someone else already investigated the issue. In such case, the build analysis tab should report the issue like so:

Author: ```suggestion
  * There's the possibility someone else has already investigated the issue. In such case, the build analysis tab should report the issue like so:
```

    ![known issue example](known-issue-example.png)
    There's no additional work required here - the bug is getting tracked and appropriate data is being collected.
  * If the error is not getting reported as a known issue and you believe it's unrelated, see the [unrelated failure](#what-to-do-if-you-determine-the-failure-is-unrelated) section for next steps.

### Option 3: The state of the main branch HEAD is bad.

* This is the very rare case where there was a build break in main, and you got unlucky. Hopefully the break has been fixed, and you want CI to rebase your change and rerun validation.
* To rebase and rerun all validation:
  * Add a comment `/azp run runtime`

>   * Or, click on "re-run all checks" in the GitHub Checks tab

Author: azp run should rebase, I am not sure of the other one

>   * Or, click on "re-run all checks" in the GitHub Checks tab

Reviewer 2: Are you sure? Do you mind if we test it tomorrow? I'm surprised, that has not been my experience.

>   * Or, click on "re-run all checks" in the GitHub Checks tab

Author: Happy to say we test it, but this is not new text, so let's not block on this one.

>   * Or, click on "re-run all checks" in the GitHub Checks tab

Reviewer 2: No need to block, but if it's not accurate, we would have to change it. Can be done later.

## What to do if you determine the failure is unrelated

An issue that's not been reported before it will look like this in the `Build Analysis` check tab:

![failed test](failed-test.png)

You can use the console log, any potential attached dumps in the artifacts section, or any other piece of information printed to help you decide if it's a regression caused by the change. Similarly, for runtime tests we will try to print the crashing stacks to aid in the investigation.

If you have considered all the diagnostic artifacts and determined the failure is definitely not caused by changes in your PR, please do this:

1. Identify a string from the logs that uniquely identifies the issue at hand. A good example of this the string `The system cannot open the device or file specified. : 'NuGet-Migrations'` for issue https://github.com/dotnet/runtime/issues/80619.

> 2. On the test failure in the tab you can select `Report repository issue`. This will prepopulate an issue with the appropriate tags and with a body similar to:

Author: I added that one. While it's useful, it's not the best way for PRs. The link does most of the work without needing to bounce around.

    ## Error Message
    Fill the error message using [known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

    ```json
    {
        "ErrorMessage": "",
        "BuildRetry": false,
        "ErrorPattern": "",
        "ExcludeConsoleLog": false
    }
    ```
    ````
    It already contains most information necessary. *The most important piece of information for you to fill out is the json blob*.

    - You can add the string you identified as uniquely identifying the issue into the `ErrorMessage` field. In case you need to use a regex, use the `ErrorPattern` field instead (this is a limitted, single-line, non-backtracking regex described [here](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#regex-matching)). This regex needs to be appropriately escaped - check the [arcade known issues](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#filling-out-known-issues-json-blob) documentation for a good guide on proper regex and JSON escaping.

>     - The field `ExcludeConsoleLog` should be `true` as much as possible. Setting it to false will mean all failures appear in the same log and since you didn't exclude the overall logs, all failures in the collection will get attributed to this one's issue bucket and errors will not be investigated properly. Due to limitations in Known Issues around rate limiting and xUnit resiliency, setting `ExcludeConsoleLog=false` is necessary in two scenarios:

Author: ```suggestion
    - The field `ExcludeConsoleLog` describes if the execution logs should be considered on top of the individual test results. **For most cases, this should be set to `true` as the failure will happen within a single test**. Setting it to `false` will mean all failures within an xUnit set of tests will also get attributed to this particular error, since there's one log describing all the problems. Due to limitations in Known Issues around rate limiting and xUnit resiliency, setting `ExcludeConsoleLog=false` is necessary in two scenarios:
```

      + Nested tests as reported to Azure DevOps. Essentially this means theory failures, which look like this when reported in Azure DevOps: ![xUnit theory seen in azure devops](theory-azdo.png).
        Adding support for this requires too many API calls, so using the console log here is necessary.
      + Native crashes in libraries also require using the console log. This is needed as the crash corrupts the test results to be reported to Azure DevOps, so only the console logs are left.
    - Optionally you can add specifics as needed like leg, configuration parameters, available dump links.

> Once the issue is open, feel free to rerun the `Build Analysis` check and the issue should be recognized as known if all was filed correctly and you are ready to merge once all unrelated issues are marked as known. However, there are some known limitations to the system as previously described. Additionally, the system only looks at the error message the stacktrace fields of an Azure DevOps test result, and the console log in the helix queue. If rerunning the check doesn't pick up the known issue and you feel it should, feel free to tag the infrastructure team for help.

Author: ```suggestion
Once the issue is open, feel free to rerun the `Build Analysis` check and the issue should be recognized as known if all was filed correctly and you are ready to merge once all unrelated issues are marked as known. However, there are some known limitations to the system as previously described. Additionally, the system only looks at the error message the stacktrace fields of an Azure DevOps test result, and the console log in the helix queue. If rerunning the check doesn't pick up the known issue and you feel it should, feel free to tag  @dotnet/runtime-infrastructure to request infrastructure team for help.
```

> After you do this, if the failure is occuring frequently as per the data captured in the recently opened issue, please disable the failing test(s) with the corresponding issue link tracking the disable in a follow-up Pull Request.

Author: ```suggestion
After you do this, if the failure is occuring frequently as per the data captured in the recently opened issue, please disable the failing test(s) with the corresponding tracking issue link in a follow-up Pull Request.
```
